File: ts/day01p2/main.ts
#!/usr/bin/env -S deno -A

export function similarityScore(left: number[], right: number[]): number {
  const lookup: { [k: number]: number } = {};
  right.map((n) => {
    lookup[n] = (lookup[n] || 0) + 1;
  });
  const score = left.reduce((acc, cur) => (acc + cur * (lookup[cur] || 0)), 0);
  return score;
}

// Learn more at https://docs.deno.com/runtime/manual/examples/module_metadata#concepts
if (import.meta.main) {
  const text = (await Deno.readTextFile("input.txt")).trimEnd();
  const left: number[] = [];
  const right: number[] = [];

  for (let line of text.split("\n")) {
    line = line.trim();
    if (line.length == 0) {
      console.log("SKIPPING");
      continue;
    }
    const [a, b] = line.split(/\s+/).map(Number);
    left.push(a);
    right.push(b);
  }
  console.log(
    `similarityScore is ${similarityScore(left, right)}`,
  );
}

File: ts/day01p2/main.py
#!/usr/bin/env python3

import sys
from collections import Counter

fname = sys.argv[1] if len(sys.argv[1:]) == 1 else "input.txt"

left, right = zip(*[map(int, line.strip().split()) for line in open(fname)])
lookup = Counter(right)
score = sum(n * lookup.get(n, 0) for n in left)
print(score)
